### ISSUE TITLE
Log before killing

### ISSUE THREAD

**Issue by essen** [ankhers/havoc#4]
Havoc should do an info level log before killing something to make it easier to understand what is being killed. The log could look like:

```
logger:info("Wrecking havoc on pid ~p~n~p~n", [Pid, process_info(Pid)]).
logger:info("Wrecking havoc on port ~p~n~p~n", [Port, erlang:port_info(Port)]).
```

**Comment by essen** [comment]
Actually on second thought having a callback just before killing would work out better. And might be good to enable disabling killing of certain pids. I'll think about it more.

**Comment by ankhers** [comment]
Resolved in 004a191d648e9b572b2465d351d8b0656b10f41d

**Comment by juhlig** [comment]
Having a callback like this is good in itself. I think it would be even better if havoc would also consider its return value, eg if it returns true kill the pid or port, if false don't. I guess that is what @essen hinted at when talking about enabling disabling killing of certain pids.

**Comment by essen** [comment]
There's undocumented options, for example `supervisors`, that make such a return value not really necessary. https://github.com/ankhers/havoc/blob/master/src/havoc.erl#L61

**Comment by ankhers** [comment]
I just updated the README to include all options. If anyone is using hex to get this package, I also just released 0.2.1 on hex.

**Comment by juhlig** [comment]
@essen I've seen those options. It is probably not necessary for `ranch` testing now, but I can imagine some use cases in a few of my own projects.

I'd argue that it allows for a more fine-grained mode of testing things than a blind rampage, even when it is restrained to a supervisor.

Also, as I understand it, the `supervisors` option is of limited use when the supervisor you are targeting is not registered. If it crashes because of the restart frequency and restarts, it will then have a pid different from the one you gave to `havoc:on`, and it won't be havoc'ed forthwith. In `ranch`, that would be the case if we wanted to target the acceptors supervisor, for example. Or, if we wanted to test that `ranch_server` stays consistent when listeners are started and stopped and crash and are restarted in a system while it is being havoc'ed, the only way to do that now is to target the `ranch` application as a whole or `ranch_sup`, both _including_  `ranch_server` that we actually wanted to examine later.

In a nutshell, I definitely see some use cases :)

@ankhers If it is decided that this feature is useful, it could be included into the `prekill_callback` facility as outlined above; in that case, this should be done soon, ie while the prekill feature is still "young"; when included later it might break or change the behavior of existing havoc suites if the callbacks there don't happen to return `true` by mere accident. Another option would be a separate callback placed before prekill which makes the decision if a pid is killable or not; that could be included later, ie when the need arises, as it won't break anything already in existence.

**Comment by essen** [comment]
Havoc is for blind rampaging though. If you need better targeting you can easily write the test case yourself. The `supervisors` option I think will target children, not the supervisor itself. The supervisor should obviously be configured to not die as a result of the rampage.

**Comment by juhlig** [comment]
> Havoc is for blind rampaging though.

Well, it doesn't _have_ to be for blind rampaging _only_. What I'm imagining here is some one-eyed rampaging, to stick to the metaphor ;) You can still go all-out full-scale blind-rampaging if you want, or rather, that's the default.

> If you need better targeting you can easily write the test case yourself.

I'd use havoc (it's there and seemingly good at what it does, no need to invent the wheel twice) to ferret misbehaviors out, zoom in on and track down the cause, fix it, and ensure it stays fixed with dedicated test cases.

That's not so easy when the component whose behavior you want to observe in terms of related components dying is itself subject to the killing. It may appear to behave when running, while really it got corrupted, was killed, restarted, and then recovered to the correct state. Or it may appear to misbehave, while really it behaved correctly while running, but recovered to a wrong state after the restart.

In the first case, the misbehavior gets ploughed under by the restart, you won't notice a problem exists. In the latter case, you do notice a problem, but will likely be searching in the wrong places for a while.

> The `supervisors` option I think will target children, not the supervisor itself. The supervisor should obviously be configured to not die as a result of the rampage.

Supervisors themselves are excluded from the killing spree, yes. Not sure if workers further down the hierarchy are targeted or not, though. Anyway, supervisors may die and be restarted if enough of their workers are killed to hit their restart limit.
